### PR TITLE
PubChem molecules containing boron and silicon

### PR DESCRIPTION
**Pull request by peastman**
This adds about 1500 molecules containing B and about 1900 containing Si.  Unfortunately OpenFF doesn't support those elements, so I can't use it to generate conformations.  Instead I use GFN-FF, which is a lot more expensive.  I'm therefore posting the script for a first round of review before I run it to generate the conformations.

### CODE REVIEW

**Comment by jchodera** [comment]
Ironically, part of the reason that OpenFF doesn't have parameters is due to the lack of good QM datasets containing these elements! I'm hoping we can also generate QCFractal datasets with the OpenFF level of theory (which is significantly cheaper) for all new datasets. (I believe we have fixed the random-retrieval ordering issue at the QCPortal level.)

If GFN-FF is very expensive, an alternative that may be attractive for this and other more exotic element datasets is to use [UFF (the Universal Force Field)](https://pubs.acs.org/doi/abs/10.1021/ja00051a040). RDKit has [implemented this force field](https://www.rdkit.org/docs/source/rdkit.Chem.rdForceFieldHelpers.html), and allows the parameters to be retrieved. A short OpenMM script could pull those parameters into an OpenMM `System` and rapidly generate simulation data the same way you do for other molecule sets. (In fact, this feature may be useful for us to include in `openmmforcefields` as a `UFFTemplateGenerator`.)

**Comment by jchodera** [comment]
Could we also add a `README` or documentation comment to this file to describe what it does?

**Comment by peastman** [comment]
GFN-FF is much more accurate the UFF, and the cost should be manageable.  We have a new 128 core node that should be able to get through everything in less than a day.  I just don't want to have to rerun it too many times, which is why I'm getting a first round of comments before running it.

**Comment by jchodera** [comment]
Understood.

* Would it be possible to include some protonation/tautomer variants in the dataset as well?
* Could you share a list of the SMILES strings that this generates for a quick manual inspection?

**Comment by peastman** [comment]
Here are the SMILES strings.

[smiles.txt](https://github.com/openmm/spice-dataset/files/11887622/smiles.txt)

**Comment by jthorton** [comment]
@peastman I would say the protocol seems great and I agree that GFN-FF is the best choice. I did notice some repeated molecules in the smiles.txt file you attached so some sort of deduplication should be performed first. 

Maybe I should put this in a sepeate issue but it might be worth considering an extra set of [boron containing molecules ](https://github.com/openforcefield/qca-dataset-submission/blob/master/submissions/2019-07-05%20OpenFF%20NCI250K%20Boron%201/optimization_inputs.pdf)@jchodera put together.

**Comment by jchodera** [comment]
Here's a sample of compounds from the `smiles.txt`:
image
And all the compounds in a PDF:
[smiles.pdf](https://github.com/openmm/spice-dataset/files/11897361/smiles.pdf)

At first glance, there are a large number of compounds where the "interesting bits" are separated by long aliphatic chains:
image
This suggests we might want to either fragment and cap these, or prioritize smaller molecules that may concentrate the "interesting" bits in smaller molecules we can sample more thoroughly.

There are also a number of really crazy molecules:
image
These are really exotic compounds that represent chemical environments unlikely to be of interest in drug discovery. 
To me, this suggests we may want to sort our dataset by increasing numbers of the element we're interested in. The compounds with one boron, for example, are much more interesting than the compounds with ten borons.

Likely protonation and tautomeric state variants of these compounds will also be very important to sample.

**Comment by peastman** [comment]
We need to think about molecules in terms of whether they're useful for learning physics.  A compound with ten borons might not be useful as a drug, but it's a gold mine for learning about how boron interacts.  Two groups separated by a flexible chain is also very useful.  It lets the groups fold over and interact noncovalently, so we get a chance to learn some different interactions.

And remember, these are all real compounds.  Every one of them is something someone thought was interesting enough to synthesize and study.  For example, that "crazy" molecule you posted?  That's [sodium tetraborate](https://pubchem.ncbi.nlm.nih.gov/compound/91932232), a.k.a Borax.  It's a common household cleaner that also happens to be toxic.  The record for it came from BindingDB which has data on how it binds to carbonic anhydrases.  It totally is a molecule people might want to simulate interacting with proteins.

**Comment by jchodera** [comment]
> And remember, these are all real compounds. Every one of them is something someone thought was interesting enough to synthesize and study. For example, that "crazy" molecule you posted? That's [sodium tetraborate](https://pubchem.ncbi.nlm.nih.gov/compound/91932232), a.k.a Borax. It's a common household cleaner that also happens to be toxic. The record for it came from BindingDB which has data on how it binds to carbonic anhydrases. It totally is a molecule people might want to simulate interacting with proteins.

That's an excellent point, though other databases show it as a [covalently dimerized form](https://pubchem.ncbi.nlm.nih.gov/compound/Sodium-Tetraborate):
image

But your point is well-taken: we do want to provide sufficient coverage of physics while also trying to ensure we have good representation of the most common potentially druglike molecules (which is what ChEMBL is supposed to contain). I'm just hoping we can balance both coverage of various chemical environments with sufficient representation of the most common chemical environments using some kind of stratification.

**Comment by peastman** [comment]
I generated the conformations.  We ended up with 174,450 conformations for 3489 molecules.

File: pubchem/createPubchemBoronSilicon.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            mol.add_conformer(state*ffunit.angstroms)
+        mol = mol.canonical_order_atoms()
+        smiles = mol.to_smiles(isomeric=True, explicit_hydrogens=True, mapped=True)
+    except:
+        print('  exception generating canonical SMILES')
+        return
+    conformations = [c.m_as(ffunit.nanometers) for c in mol.conformers]
+    conformations = [c-np.average(c, axis=0) for c in conformations]
+    group = outputfile.create_group(name)
+    group.create_dataset('smiles', data=[smiles], dtype=h5py.string_dtype())
```
Comment (jchodera) [inline pubchem/createPubchemBoronSilicon.py:99]: Could `compression` be used here to save space?
https://docs.h5py.org/en/stable/high/dataset.html#filter-pipeline

**Comment by jchodera** [comment]
This looks good! I'm just worried about committing a 70MB HDF5 file. I suggested a couple of small changes that might reduce file size without reducing quality.

**Comment by peastman** [comment]
The same directory already contains six other files that are over 50 MB each.  I think it's best to keep everything in the same format.  The files are big, but there aren't that many of them.  Github requires that individual files be under 100 MB, and they "strongly recommend" that the total size of the repository stay under 5 GB.  I don't think we're in any danger of hitting that limit for a very long time.

**Comment by jchodera** [comment]
What about at least enabling compression? That should be transparent.

**Comment by peastman** [comment]
I could try that on future datasets.  I'd prefer not to have to regenerate this one.  Note that git automatically stores files in compressed form.  So compressing the individual files would make the checked out files in the working copy smaller, but it wouldn't significantly affect the size of the repository itself.

**Pull request merged**